### PR TITLE
Remove polyfills for PHP < 5.6

### PR DESCRIPTION
**Pull request by mattyrob**
## Description
This PR will fix issue #451 
It contains backports, as follows:
Ticket: 47698
Commits:
45636 
45637
45638

Ticket: 47699
Commits:
46205
46206
46208
46377
46455
46464
46560 

## Motivation and context
The core code contains several functions for backwards compatibility purposes that are now part of PHP since the minimum support version has been increased. As such are no longer needed in the core code.

## How has this been tested?
Upstream ticket has not impacted on unit tests and has been a core part of WordPress for about a year.

## Types of changes
Bug Fix
Code modernisation

### CODE REVIEW

File: src/wp-admin/includes/update-core.php [hunk 1]
```diff
@@ -225,6 +225,19 @@ function update_core($from, $to) {
 	elseif ( !$mysql_compat )
 		return new WP_Error( 'mysql_not_compatible', sprintf( __('The update cannot be installed because ClassicPress %1$s requires MySQL version %2$s or higher. You are running version %3$s.'), $wp_version, $required_mysql_version, $mysql_version ) );
 
+	// Add a warning when the JSON PHP extension is missing.
+	if ( ! extension_loaded( 'json' ) ) {
+		return new WP_Error(
+			'php_not_compatible_json',
+			sprintf(
+				/* translators: 1: WordPress version number, 2: The PHP extension name needed. */
+				__( 'The update cannot be installed because WordPress %1$s requires the %2$s PHP extension.' ),
```
Comment (bahiirwa) [inline src/wp-admin/includes/update-core.php:234]: Replace WordPress with CP

**Comment by bahiirwa** [comment]
This is nice for the code. However `WP-r46205: Code Modernization: Deprecate the `Services_JSON` and `Services_JSON_Error` classes.` had a couple of issues - ref:: https://core.trac.wordpress.org/ticket/47699#comment:6

How do we resolve that since we don't have site health?

File: src/wp-admin/includes/update-core.php (hunk 1 above)
Comment (mattyrob) [reply src/wp-admin/includes/update-core.php]: Thanks for spotting that, update to follow.

**Comment by mattyrob** [comment]
> This is nice for the code. However `` WP-r46205: Code Modernization: Deprecate the `Services_JSON` and `Services_JSON_Error` classes. `` had a couple of issues - ref:: https://core.trac.wordpress.org/ticket/47699#comment:6
> 
> How do we resolve that since we don't have site health?

Looking at the thread it seems like a very low proportion of servers would be affected.

Reviewed the comments about the code and seeing a plugin I use in the list too it seems that most of the popular plugins and themes are using these files as a fall back for when the server PHP functions do not exist.

Perhaps this should go into ClassicPress 2.0 as a possible breaking change.

**Comment by nylen** [comment]
Note, need to remove WP-r46205 from this PR since it would be a breaking change, even if for an extremely small number of sites.

File: src/wp-includes/class-json.php [hunk 2]
```diff
@@ -1,4 +1,6 @@
 <?php
+_deprecated_file( basename( __FILE__ ), '5.3.0', null, 'The PHP native JSON extension is now a requirement.' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:2]:
```suggestion
_deprecated_file( basename( __FILE__ ), 'WP-5.3.0', null, 'The PHP native JSON extension is now a requirement.' );
```

File: src/wp-includes/class-json.php [hunk 3]
```diff
@@ -122,6 +124,8 @@ class Services_JSON
    /**
     * constructs a new JSON instance
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:127]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 4]
```diff
@@ -142,6 +146,8 @@ class Services_JSON
     */
     function __construct( $use = 0 )
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:149]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 5]
```diff
@@ -150,8 +156,13 @@ function __construct( $use = 0 )
 
 	/**
 	 * PHP4 constructor.
+     *
+     * @deprecated 5.3.0 Use __construct() instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:160]:
```suggestion
     * @deprecated WP-5.3.0 Use __construct() instead.
```

File: src/wp-includes/class-json.php [hunk 6]
```diff
@@ -150,8 +156,13 @@ function __construct( $use = 0 )
...
+     *
+     * @see __construct()
 	 */
 	public function Services_JSON( $use = 0 ) {
+		_deprecated_constructor( 'Services_JSON', '5.3.0', get_class( $this ) );
```
Comment (nylen) [inline src/wp-includes/class-json.php:165]:
```suggestion
		_deprecated_constructor( 'Services_JSON', 'WP-5.3.0', get_class( $this ) );
```

File: src/wp-includes/class-json.php [hunk 7]
```diff
@@ -166,12 +177,16 @@ public function Services_JSON( $use = 0 ) {
     * provides a slower PHP-only method for installations
     * that lack the multibye string extension.
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:180]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 8]
```diff
@@ -166,12 +177,16 @@ public function Services_JSON( $use = 0 ) {
...
+    *
     * @param    string  $utf16  UTF-16 character
     * @return   string  UTF-8 character
     * @access   private
     */
     function utf162utf8($utf16)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:188]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 9]
```diff
@@ -210,12 +225,16 @@ function utf162utf8($utf16)
     * provides a slower PHP-only method for installations
     * that lack the multibye string extension.
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:228]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 10]
```diff
@@ -210,12 +225,16 @@ function utf162utf8($utf16)
...
+    *
     * @param    string  $utf8   UTF-8 character
     * @return   string  UTF-16 character
     * @access   private
     */
     function utf82utf16($utf8)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:236]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 11]
```diff
@@ -250,6 +269,8 @@ function utf82utf16($utf8)
    /**
     * encodes an arbitrary variable into JSON format (and sends JSON Header)
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:272]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 12]
```diff
@@ -260,12 +281,16 @@ function utf82utf16($utf8)
     */
     function encode($var)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:284]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 13]
```diff
@@ -260,12 +281,16 @@ function utf82utf16($utf8)
...
+
         header('Content-type: application/json');
         return $this->encodeUnsafe($var);
     }
     /**
     * encodes an arbitrary variable into JSON format without JSON Header - warning - may allow XSS!!!!)
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:292]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 14]
```diff
@@ -276,6 +301,8 @@ function encode($var)
     */
     function encodeUnsafe($var)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:304]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 15]
```diff
@@ -287,6 +314,8 @@ function encodeUnsafe($var)
     /**
     * PRIVATE CODE that does the work of encodes an arbitrary variable into JSON format
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:317]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 16]
```diff
@@ -297,6 +326,7 @@ function encodeUnsafe($var)
     */
     function _encode($var)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:329]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 17]
```diff
@@ -531,6 +561,8 @@ function _encode($var)
    /**
     * array-walking function for use in generating JSON-formatted name-value pairs
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:564]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 18]
```diff
@@ -539,6 +571,8 @@ function _encode($var)
     */
     function name_value($name, $value)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:574]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 19]
```diff
@@ -551,13 +585,17 @@ function name_value($name, $value)
    /**
     * reduce a string by removing leading and trailing comments and whitespace
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:588]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 20]
```diff
@@ -551,13 +585,17 @@ function name_value($name, $value)
...
+    *
     * @param    $str    string      string value to strip of comments and whitespace
     *
     * @return   string  string value stripped of comments and whitespace
     * @access   private
     */
     function reduce_string($str)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:597]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 21]
```diff
@@ -578,6 +616,8 @@ function reduce_string($str)
    /**
     * decodes a JSON string into appropriate variable
     *
+    * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:619]:
```suggestion
    * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 22]
```diff
@@ -589,6 +629,8 @@ function reduce_string($str)
     */
     function decode($str)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:632]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 23]
```diff
@@ -774,7 +816,7 @@ function decode($str)
                                 $parts = array();
 
                                if (preg_match('/^\s*(["\'].*[^\\\]["\'])\s*:/Uis', $slice, $parts)) {
- 	                              // "name":value pair
+                                    // "name":value pair
```
Comment (nylen) [inline src/wp-includes/class-json.php:819]: Don't change the whitespace here:

```suggestion
 	                              // "name":value pair
```

File: src/wp-includes/class-json.php [hunk 24]
```diff
@@ -866,10 +908,14 @@ function decode($str)
     }
 
     /**
+     * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:911]:
```suggestion
     * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 25]
```diff
@@ -866,10 +908,14 @@ function decode($str)
...
+     *
      * @todo Ultimately, this should just call PEAR::isError()
      */
     function isError($data, $code = null)
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:917]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 26]
```diff
@@ -881,27 +927,37 @@ function isError($data, $code = null)
...
+     * Calculates length of string in bytes
+     *
+     * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
+     *
+     * @param string
+     * @return integer length
+     */
     function strlen8( $str )
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:939]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 27]
```diff
@@ -881,27 +927,37 @@ function isError($data, $code = null)
...
+     * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
+     *
+     * @param string
+     * @param integer start
+     * @param integer length
+     * @return integer length
+     */
     function substr8( $string, $start, $length=false )
     {
+        _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:959]:
```suggestion
        _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 28]
```diff
@@ -917,17 +973,31 @@ function substr8( $string, $start, $length=false )
 
     class Services_JSON_Error extends PEAR_Error
     {
+	    /**
+	     * PHP5 constructor.
+	     *
+	     * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:979]:
```suggestion
	     * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 29]
```diff
@@ -917,17 +973,31 @@ function substr8( $string, $start, $length=false )
...
+	     */
         function __construct($message = 'unknown error', $code = null,
                                      $mode = null, $options = null, $userinfo = null)
         {
+            _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:984]:
```suggestion
            _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 30]
```diff
@@ -917,17 +973,31 @@ function substr8( $string, $start, $length=false )
...
+
             parent::PEAR_Error($message, $code, $mode, $options, $userinfo);
         }
 
-	public function Services_JSON_Error($message = 'unknown error', $code = null,
+	    /**
+	     * PHP4 constructor.
+	     *
+	     * @deprecated 5.3.0 Use __construct() instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:992]:
```suggestion
	     * @deprecated WP-5.3.0 Use __construct() instead.
```

File: src/wp-includes/class-json.php [hunk 31]
```diff
@@ -917,17 +973,31 @@ function substr8( $string, $start, $length=false )
...
+	     *
+	     * @see __construct()
+	     */
+		public function Services_JSON_Error($message = 'unknown error', $code = null,
                                      $mode = null, $options = null, $userinfo = null) {
-		self::__construct($message = 'unknown error', $code = null,
-                                     $mode = null, $options = null, $userinfo = null);
-	}
+			_deprecated_constructor( 'Services_JSON_Error', '5.3.0', get_class( $this ) );
```
Comment (nylen) [inline src/wp-includes/class-json.php:998]:
```suggestion
			_deprecated_constructor( 'Services_JSON_Error', 'WP-5.3.0', get_class( $this ) );
```

File: src/wp-includes/class-json.php [hunk 32]
```diff
@@ -939,18 +1009,25 @@ class Services_JSON_Error
     {
 	    /**
 	     * PHP5 constructor.
+	     *
+	     * @deprecated 5.3.0 Use the PHP native JSON extension instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:1013]:
```suggestion
	     * @deprecated WP-5.3.0 Use the PHP native JSON extension instead.
```

File: src/wp-includes/class-json.php [hunk 33]
```diff
@@ -939,18 +1009,25 @@ class Services_JSON_Error
...
 	     */
         function __construct( $message = 'unknown error', $code = null,
                                      $mode = null, $options = null, $userinfo = null )
         {
-
+            _deprecated_function( __METHOD__, '5.3.0', 'The PHP native JSON extension' );
```
Comment (nylen) [inline src/wp-includes/class-json.php:1018]:
```suggestion
            _deprecated_function( __METHOD__, 'WP-5.3.0', 'The PHP native JSON extension' );
```

File: src/wp-includes/class-json.php [hunk 34]
```diff
@@ -939,18 +1009,25 @@ class Services_JSON_Error
...
         }
 
 	    /**
 	     * PHP4 constructor.
+	     *
+	     * @deprecated 5.3.0 Use __construct() instead.
```
Comment (nylen) [inline src/wp-includes/class-json.php:1024]:
```suggestion
	     * @deprecated WP-5.3.0 Use __construct() instead.
```

File: src/wp-includes/class-json.php [hunk 35]
```diff
@@ -939,18 +1009,25 @@ class Services_JSON_Error
...
+	     *
+	     * @see __construct()
 	     */
 		public function Services_JSON_Error( $message = 'unknown error', $code = null,
 	                                     $mode = null, $options = null, $userinfo = null ) {
+			_deprecated_constructor( 'Services_JSON_Error', '5.3.0', get_class( $this ) );
```
Comment (nylen) [inline src/wp-includes/class-json.php:1030]:
```suggestion
			_deprecated_constructor( 'Services_JSON_Error', 'WP-5.3.0', get_class( $this ) );
```

File: src/wp-includes/spl-autoload-compat.php [hunk 36]
```diff
@@ -5,96 +5,10 @@
  *
  * See https://core.trac.wordpress.org/ticket/41134
  *
+ * @deprecated 5.3.0 No longer needed as the minimum PHP requirement has moved beyond PHP 5.3.
```
Comment (nylen) [inline src/wp-includes/spl-autoload-compat.php:8]:
```suggestion
 * @deprecated WP-5.3.0 No longer needed as the minimum PHP requirement has moved beyond PHP 5.3.
```

File: src/wp-includes/spl-autoload-compat.php [hunk 37]
```diff
@@ -5,96 +5,10 @@
...
-	 *
-	 * @since WP-4.6.0
-	 *
-	 * @return array List of autoloader functions.
-	 */
-	function spl_autoload_functions() {
-		return $GLOBALS['_wp_spl_autoloaders'];
-	}
-}
+_deprecated_file( basename( __FILE__ ), '5.3.0', null, 'SPL can no longer be disabled as of PHP 5.3.' );
```
Comment (nylen) [inline src/wp-includes/spl-autoload-compat.php:14]:
```suggestion
_deprecated_file( basename( __FILE__ ), 'WP-5.3.0', null, 'SPL can no longer be disabled as of PHP 5.3.' );
```

**Review by nylen — APPROVED** [review]
Assuming the tests pass, this is good to go.

**Pull request merged**